packages/next/lib/recursive-delete.ts

```diff
@@ -0,0 +1,38 @@
+import fs from 'fs'
+import { join } from 'path'
+import { promisify } from 'util'
+
+const readdir = promisify(fs.readdir)
+const stat = promisify(fs.stat)
+const rmdir = promisify(fs.rmdir)
+const unlink = promisify(fs.unlink)
+
+/**
+ * Recursively delete directory contents
+ * @param  {string} dir Directory to delete the contents of
+ * @param  {RegExp} filter Filter for the file name, only the name part is considered, not the full path
+ * @returns Promise void
+ */
+export async function recursiveDelete(dir: string, filter?: RegExp, ensure?: boolean): Promise<void> {
+  let result
+  try {
+    result = await readdir(dir)
+  } catch (e) {
+    if (ensure) throw e
+    // If the dir does not exist we dont need to do anything
```

Comment:
Shouldn't we be checking for the proper code to assume this doesn't exist?

Reply:
Another potential API _could_ be to return something like a `Promise<number>` that reflects the amount of files/folders that were deleted. Consumers of this function could then do this `ensure` check themselves instead. This way the function stays more scoped to its core intention.

Reply:
"Using fs.access() to check for the accessibility of a file before calling fs.open(), fs.readFile() or fs.writeFile() is not recommended. Doing so introduces a race condition, since other processes may change the file's state between the two calls. Instead, user code should open/read/write the file directly and handle the error raised if the file is not accessible."

-Node's fs docs

Reply:
@janpot you'd have to count those files first. It doesn't matter how many files we deleted all that matters is that folder is empty. Ensure is only for checking if the *initial* directory exists. Deeper than the initial directory, it'll just catch errors because it's only reading directories that'vd been `fs.stat` already

Reply:
You don't have to count first, you can just pass a `count: number` around and increment each time you call `unlink`. But, yeah, it was only a thought.

Reply:
@janpot so it'd return a number of files it deleted that you'd compare to what? Just to say how many were deleted? If you wanted to verify that number was correct and it didn't miss any, you'd have to read and count all the files before calling the delete function

Reply:
> you'd compare to what?

If you want to figure out if something or nothing is deleted, I guess you'd compare it to `0`

> you'd have to read and count all the files before calling the delete function

Not really, at the time you call the `unlink`, you've already verified that the file you're deleting exists, so it's just a matter of keeping count of how many times you call that `unlink` function. But again, it doesn't matter, API design is in big part subjective and I don't necessarily claim it's the better way to do it. I just meant to bring up an alternative, just as inspiration. I regret bringing it up.

Reply:
@janpot we just don't have a use case right now and could always implement something like that when we need it. Keeping it as simple as possible for now